File: scripts/check-ai-recommendation.ts
/**
 * Check AI recommendation layer for Al Ghadeer.
 *
 * Always validates deterministic fallback (must pass).
 * Optionally tests OpenRouter when OPENROUTER_API_KEY is set.
 * Never prints API keys.
 *
 * Run: npm run check:ai
 */

import { existsSync, readFileSync } from "fs";
import { join, resolve } from "path";
import { fallbackRecommendation } from "../lib/fallbackRecommendation";
import { generateRecommendation } from "../lib/aiRecommendation";
import type { CommunityGapOutputs, DistrictRecommendation } from "../lib/communityGapTypes";

const ROOT = resolve(__dirname, "..");
const DATA_PATH = join(ROOT, "processed", "community_gap_outputs.json");
const DISTRICT_NAME = "Al Ghadeer";

const REQUIRED_KEYS: (keyof DistrictRecommendation)[] = [
  "district_summary",
  "main_gap",
  "recommended_intervention",
  "why_this_matters",
  "confidence_note",
  "uncertainty_note",
];

function loadEnvFiles(): void {
  const candidates = [
    join(ROOT, ".env.local"),
    join(ROOT, ".env"),
    join(ROOT, "..", ".env"),
  ];

  for (const filePath of candidates) {
    if (!existsSync(filePath)) continue;
    const lines = readFileSync(filePath, "utf8").split("\n");
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#") || !trimmed.includes("=")) continue;
      const eq = trimmed.indexOf("=");
      const key = trimmed.slice(0, eq).trim();
      let value = trimmed.slice(eq + 1).trim();
      if (
        (value.startsWith('"') && value.endsWith('"')) ||
        (value.startsWith("'") && value.endsWith("'"))
      ) {
        value = value.slice(1, -1);
      }
      if (key && process.env[key] === undefined) {
        process.env[key] = value;
      }
    }
  }
}

function validateOutput(rec: DistrictRecommendation, label: string): void {
  const missing = REQUIRED_KEYS.filter((k) => !(k in rec));
  if (missing.length > 0) {
    throw new Error(`${label}: missing keys: ${missing.join(", ")}`);
  }

  const empty = REQUIRED_KEYS.filter((k) => !rec[k]?.trim());
  if (empty.length > 0) {
    throw new Error(`${label}: empty fields: ${empty.join(", ")}`);
  }
}

function printResult(rec: DistrictRecommendation, source: string): void {
  console.log(`\n=== ${DISTRICT_NAME} (${source}) ===`);
  console.log(JSON.stringify(rec, null, 2));
}

async function main(): Promise<number> {
  loadEnvFiles();

  if (!existsSync(DATA_PATH)) {
    console.error(`ERROR: missing ${DATA_PATH}`);
    return 1;
  }

  const payload = JSON.parse(
    readFileSync(DATA_PATH, "utf8")
  ) as CommunityGapOutputs;

  const district = payload.districts.find((d) => d.district === DISTRICT_NAME);
  if (!district) {
    console.error(`ERROR: district not found in JSON: ${DISTRICT_NAME}`);
    return 1;
  }

  console.log(`Loaded community_gap_outputs.json — checking ${DISTRICT_NAME}`);

  // 1. Fallback (required)
  const fallbackRec = fallbackRecommendation(district);
  validateOutput(fallbackRec, "fallback");
  printResult(fallbackRec, "fallback");

  if (
    fallbackRec.recommended_intervention !==
    district.classification.recommended_intervention_category
  ) {
    throw new Error("fallback recommended_intervention does not match pipeline");
  }

  if (
    !fallbackRec.district_summary
      .toLowerCase()
      .includes("top priority in the current dataset")
  ) {
    throw new Error("Al Ghadeer fallback missing top-priority wording");
  }

  console.log("\nFallback check: PASSED");

  // 2. OpenRouter (optional)
  const apiKey = process.env.OPENROUTER_API_KEY?.trim();
  if (apiKey) {
    console.log("\nOPENROUTER_API_KEY found — testing OpenRouter path...");
    try {
      const { recommendation, source } = await generateRecommendation(district);
      validateOutput(recommendation, `openrouter (${source})`);
      printResult(recommendation, source);
      if (source === "llm") {
        console.log("\nOpenRouter check: PASSED (llm)");
      } else {
        console.log(
          "\nOpenRouter check: PASSED (fell back to deterministic output)"
        );
      }
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      console.log(`\nOpenRouter check: SKIPPED (${message})`);
      console.log("Fallback still passed — overall check OK.");
    }
  } else {
    console.log("\nOPENROUTER_API_KEY not set — skipping OpenRouter test.");
  }

  console.log("\nAll required checks passed.");
  return 0;
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error("CHECK FAILED:", err instanceof Error ? err.message : err);
    process.exit(1);
  });

File: scripts/check_ai_recommendation.py
#!/usr/bin/env python3
"""
Check AI recommendation layer for Al Ghadeer.

Always validates deterministic fallback (must pass).
Optionally tests OpenRouter when OPENROUTER_API_KEY is set.
Never prints API keys.
"""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from ai.fallback import fallback_recommendation  # noqa: E402
from ai.recommendation import generate_recommendation  # noqa: E402

REQUIRED_KEYS = (
    "district_summary",
    "main_gap",
    "recommended_intervention",
    "why_this_matters",
    "confidence_note",
    "uncertainty_note",
)

DISTRICT_NAME = "Al Ghadeer"
DATA_PATH = ROOT / "processed" / "community_gap_outputs.json"


def _load_env_files() -> None:
    """Load .env files into os.environ without printing values."""
    candidates = [
        ROOT / ".env.local",
        ROOT / ".env",
        ROOT.parent / ".env",
    ]
    for path in candidates:
        if not path.is_file():
            continue
        for line in path.read_text(encoding="utf-8").splitlines():
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, _, value = line.partition("=")
            key = key.strip()
            value = value.strip().strip('"').strip("'")
            if key and key not in os.environ:
                os.environ[key] = value


def _validate_output(rec: dict[str, str], label: str) -> None:
    missing = [k for k in REQUIRED_KEYS if k not in rec]
    if missing:
        raise AssertionError(f"{label}: missing keys: {', '.join(missing)}")

    empty = [k for k in REQUIRED_KEYS if not str(rec[k]).strip()]
    if empty:
        raise AssertionError(f"{label}: empty fields: {', '.join(empty)}")


def _print_result(rec: dict[str, str], source: str) -> None:
    print(f"\n=== {DISTRICT_NAME} ({source}) ===")
    print(json.dumps(rec, indent=2, ensure_ascii=False))


def main() -> int:
    _load_env_files()

    if not DATA_PATH.is_file():
        print(f"ERROR: missing {DATA_PATH}", file=sys.stderr)
        return 1

    payload = json.loads(DATA_PATH.read_text(encoding="utf-8"))
    district = next(
        (d for d in payload["districts"] if d["district"] == DISTRICT_NAME),
        None,
    )
    if district is None:
        print(f"ERROR: district not found in JSON: {DISTRICT_NAME}", file=sys.stderr)
        return 1

    print(f"Loaded {DATA_PATH.name} — checking {DISTRICT_NAME}")

    # 1. Fallback (required)
    fallback_rec = fallback_recommendation(district)
    _validate_output(fallback_rec, "fallback")
    _print_result(fallback_rec, "fallback")

    assert (
        fallback_rec["recommended_intervention"]
        == district["classification"]["recommended_intervention_category"]
    )
    assert "top priority in the current dataset" in fallback_rec["district_summary"].lower()

    print("\nFallback check: PASSED")

    # 2. OpenRouter (optional)
    api_key = os.environ.get("OPENROUTER_API_KEY", "").strip()
    if api_key:
        print("\nOPENROUTER_API_KEY found — testing OpenRouter path...")
        try:
            result = generate_recommendation(district, force_fallback=False)
            rec = result["recommendation"]
            source = result["source"]
            _validate_output(rec, f"openrouter ({source})")
            _print_result(rec, source)
            if source == "llm":
                print("\nOpenRouter check: PASSED (llm)")
            else:
                print("\nOpenRouter check: PASSED (fell back to deterministic output)")
        except Exception as exc:
            print(f"\nOpenRouter check: SKIPPED ({exc})")
            print("Fallback still passed — overall check OK.")
    else:
        print("\nOPENROUTER_API_KEY not set — skipping OpenRouter test.")

    print("\nAll required checks passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
